File: backend/app/models/technical.py
"""技术响应表数据模型。

对应 SQLite 表 technical_responses，用于 product_spec 类要求的
"标书要求值 vs 实际值"对比管理。
"""

from typing import Optional

from pydantic import BaseModel


class TechnicalResponse(BaseModel):
    """技术响应模型，对应 technical_responses 表。

    Attributes:
        id: 主键，自增。
        tender_id: 外键，关联 tenders.id。
        requirement_id: 外键，关联 tender_requirements.id。
        spec_name: 参数名称（从 requirement.title 派生）。
        required_value: 标书要求值（从 requirement content/numeric 派生）。
        actual_value: 实际值（用户手填）。
        is_hard: 是否硬性指标。
        remark: 备注。
        created_at: 记录创建时间（ISO 8601）。
        updated_at: 记录更新时间（ISO 8601）。
    """

    id: Optional[int] = None
    tender_id: int = 0
    requirement_id: int = 0
    spec_name: Optional[str] = None
    required_value: Optional[str] = None
    actual_value: Optional[str] = None
    response_status: str = "pending"
    is_hard: bool = True
    remark: Optional[str] = None
    created_at: str = ""
    updated_at: str = ""

File: frontend/src/api/technical.ts
import client from './client';
import type {
  TechnicalResponse,
  TechnicalResponseInput,
  TechnicalResponseBatchItem,
} from '../types';

/**
 * Technical Response API module.
 */

/** GET /tenders/{id}/technical — fetch technical responses (auto-init on first access). */
export async function getTechnicalResponses(
  tenderId: number,
): Promise<TechnicalResponse[]> {
  const res = await client.get<unknown[]>(`/tenders/${tenderId}/technical`);
  return res.data.map((raw) => normalizeTechnicalResponse(raw));
}

/** PUT /tenders/{id}/technical/{responseId} — update a single technical response. */
export async function updateTechnicalResponse(
  tenderId: number,
  responseId: number,
  data: TechnicalResponseInput,
): Promise<TechnicalResponse> {
  const res = await client.put<unknown>(
    `/tenders/${tenderId}/technical/${responseId}`,
    data,
  );
  return normalizeTechnicalResponse(res.data);
}

/** PUT /tenders/{id}/technical/batch — batch update technical responses. */
export async function batchUpdateTechnicalResponses(
  tenderId: number,
  items: TechnicalResponseBatchItem[],
): Promise<TechnicalResponse[]> {
  const res = await client.put<unknown[]>(
    `/tenders/${tenderId}/technical/batch`,
    { items },
  );
  return res.data.map((raw) => normalizeTechnicalResponse(raw));
}

/** Normalize raw API response to TechnicalResponse. */
function normalizeTechnicalResponse(raw: unknown): TechnicalResponse {
  const item = raw as Record<string, unknown>;
  const status = item.response_status;
  return {
    id: Number(item.id),
    tender_id: Number(item.tender_id),
    requirement_id: Number(item.requirement_id),
    spec_name: item.spec_name == null ? null : String(item.spec_name),
    required_value:
      item.required_value == null ? null : String(item.required_value),
    actual_value:
      item.actual_value == null ? null : String(item.actual_value),
    response_status:
      status === 'met' || status === 'deviated' || status === 'superior'
        ? status
        : 'pending',
    is_hard: Boolean(item.is_hard),
    remark: item.remark == null ? null : String(item.remark),
    created_at: String(item.created_at || ''),
    updated_at: String(item.updated_at || ''),
  };
}
